### PR TITLE
Feat 31 add calendar component

### PR DESCRIPTION
**Pull request by topherfangio**
## Context
This PR introduces the new Daisy Cally component, a date picker input with enhanced floating label support. The implementation includes both the core calendar component and an input component that works seamlessly with it, following DaisyUI 5 design patterns.

## Related Issues
- Fixes [Add new DaisyUI 5 Calendar component #31](https://github.com/profoundry-us/loco_motion/issues/31)

## Type of Change
- New feature (non-breaking change which adds functionality)
- Component update/addition

### New Components
- CallyComponent: A reusable calendar component with month navigation
- CallyInputComponent: A text input with integrated calendar picker
- Stimulus Controller: Handles the interaction between input and calendar

### Key Features
- Floating Label Support: Uses zero-width spaces for smooth label transitions without flicker
- Keyboard Navigation: Full keyboard support for accessibility (arrow keys, space, enter)
- Responsive Design: Works across different screen sizes with auto-scrolling to keep the calendar in view
- Customizable: Supports custom icons, date formats, and styling
- Integrates with the LabelableComponent concern for consistent form handling

### Technical Details
- Uses HTML's native popover API for better performance and accessibility
- Implements Stimulus.js for interactive behavior with comprehensive event handling
- Follows DaisyUI 5 design patterns with built-in border styling
- Includes comprehensive test coverage
- Maintains focus appropriately when navigating between input and calendar
- Properly handles edge cases like empty inputs with floating labels

### Other Changes
- Fixed issues with figures not showing captions
- Fixed fieldset textarea rendering issues
- Added error modifier to doc_note component for improved documentation
- Fixed typos in select documentation
- Added check in set_loco_parent_patch to ensure we catch nil references
- Used shorthand floating_placeholder for label examples
- Fixed bug in boxed countdown example
- Added todos about better API doc buttons

### Screenshots
Screenshot 2025-05-17 at 5 50 06 PM

Screenshot 2025-05-17 at 5 49 50 PM

### CODE REVIEW

File: .windsurf/workflows/new_component_workflow.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+**Content template**:
+
+```ruby
+module Daisy
+  module {ModuleName}
+    class {ClassName}Component < LocoMotion::BaseComponent
+      self.component_name = "{component_name}"
+
+      # Define parts here (empty by default)
+      # part :label
```
Comment (topherfangio) [inline .windsurf/workflows/new_component_workflow.md:36]: Should say define_part

Comment (topherfangio) [reply .windsurf/workflows/new_component_workflow.md:36]: Maybe be use (e.g.) syntax

File: app/components/daisy/actions/dropdown_component.html.haml [hunk 2]
```diff
@@ -1,6 +1,8 @@
 = part(:component) do
   - if activator?
     = activator
+  - elsif input?
```
Comment (topherfangio) [inline app/components/daisy/actions/dropdown_component.html.haml:4]: See if we can remove this in favor of a regular input with a popover

File: app/components/daisy/actions/dropdown_component.rb [hunk 3]
```diff
@@ -16,6 +16,8 @@
 #   and is styled automatically.
 # @slot activator A custom (i.e. non-button) activator for the dropdown.
 #   Automatically adds the `role="button"` and `tabindex="0"` attributes.
+# @slot input A text input field that can be included in the dropdown menu, useful
```
Comment (topherfangio) [inline app/components/daisy/actions/dropdown_component.rb:19]: These docs are wrong and the whole slot can probably be removed... or actually do what it says and put the input in the dropdown menu

File: app/components/daisy/data_input/cally_component.rb [hunk 4]
```diff
@@ -0,0 +1,180 @@
+# frozen_string_literal: true
+
+module Daisy
+  module DataInput
+    # The Cally component provides a customizable calendar interface for date selection.
+    # It supports both single date and date range selection, with configurable display
+    # options including the number of months to show and navigation controls.
+    #
+    # @part component The root calendar element that contains all other parts.
+    # @part months The container element that holds the month components.
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_component.rb:10]: Individual month

File: app/components/daisy/data_input/cally_component.rb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+      # handles navigation between months and can display multiple months.
+      #
+      # @param change [String] ID of an input to update with the selected date
+      # @param update [String] ID of an element to update with the selected date
+      # @param id [String] The ID of the calendar element
+      # @param value [String, Date] The currently selected date or range
+      # @param min [String, Date] The minimum selectable date
+      # @param max [String, Date] The maximum selectable date
+      # @param today [String, Date] The date to consider as 'today'
+      # @param months [Integer] Number of months to display (default: 1)
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_component.rb:103]: Defaults to 1

File: app/components/daisy/data_input/cally_component.rb [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      define_parts :months
+
+      # Initializes a new CallyComponent.
+      #
+      # The Cally component provides a customizable calendar interface for date
+      # selection.  It supports single date selection by default and can be
+      # configured for date range selection.  The component automatically
+      # handles navigation between months and can display multiple months.
+      #
+      # @param change [String] ID of an input to update with the selected date
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_component.rb:96]: Should note that these are mutually exclusive

File: app/components/daisy/data_input/cally_input_component.rb [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  module DataInput
+    # A specialized input component that combines a text input with a calendar picker.
+    # The calendar appears in a popover when the input is focused or clicked.
+    #
+    # @part popover The container for the calendar popover.
+    #
+    # @slot input The text input component.
+    # @slot calendar The calendar component that appears in the popover.
+    #
+    # @option id [String] A unique identifier for the input (default: auto-generated).
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_input_component.rb:11]: I think these all need kws and a @param at the top

File: app/components/daisy/data_input/cally_input_controller.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    if (event.relatedTarget && this.calendarTarget.contains(event.relatedTarget)) {
+      return
+    }
+
+    this.popoverTarget.togglePopover(false)
+  }
+
+  /**
+   * Handles the popover toggle event.
+   * - When opening: Sets a space in empty inputs to prevent floating label flicker
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_input_controller.js:83]: Mention zero width

File: app/components/daisy/data_input/cally_input_controller.js [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    //
+    else if (isOpen && (event.key === 'Enter' || event.code === 'Enter')) {
+      // Stop the enter key from triggering the form submission
+      event.preventDefault();
+
+      this.popoverTarget.togglePopover(false)
+    }
+
+    //
+    // ARROW DOWN / ARROW UP - Opens the popover and focuses the calendar
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_input_controller.js:153]: See if we need to remove this for accessibility of page scroll on store keys. You can currently focus an input and not see it's calendar. Maybe we can scroll into view on open.

File: app/components/daisy/data_input/cally_input_controller.js [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    //
+    else if ((event.key === 'ArrowDown' || event.code === 'ArrowUp') && !hasModifierKeys) {
+      event.preventDefault()
+
+      // If the calendar is not already open, open it
+      if (!this.popoverTarget.matches(':popover-open')) {
```
Comment (topherfangio) [inline app/components/daisy/data_input/cally_input_controller.js:159]: Extract this into an isOpen variable

File: app/components/daisy/data_input/text_input_component.rb [hunk 11]
```diff
@@ -73,8 +73,6 @@ class Daisy::DataInput::TextInputComponent < LocoMotion::BaseComponent
   #
   # @option kws value [String] The initial value of the text input.
   #
-  # @option kws placeholder [String] Placeholder text for the text input.
-  #
   # @option kws type [String] The type of input (text, password, email, etc.).
   #   Defaults to "text".
   #
```
Comment (topherfangio) [inline app/components/daisy/data_input/text_input_component.rb:78]: Document the new change option and add to changelog

File: docs/demo/app/helpers/application_helper.rb [hunk 12]
```diff
@@ -28,4 +28,22 @@ def doc_note(*args, **kws, &block)
...
   end
+
+  # Creates a link to a component example page with consistent styling
+  #
+  # @param text [String] The button text
+  # @param component_class [String] The full component class name (e.g., "Daisy::DataInput::TextInputComponent")
+  # @param size [String] The button size (default: "btn-xs")
+  # @param html_options [Hash] Additional HTML options for the button
+  # @return [String] HTML for the component link button
+  def component_link(text, component_class, css: "btn-xs", **options)
```
Comment (topherfangio) [inline docs/demo/app/helpers/application_helper.rb:39]: We should probably make this a doc component

File: docs/demo/app/views/examples/daisy/data_input/calendars.html.haml [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      but is specialized for updating an input element's `value`.
+
+    = doc_code(language: "js") do
+      :plain
+        { onchange: "document.getElementById('\#{@update}').value = this.value" }
+
+    = doc_note(modifier: :tip, css: "mb-8") do
+      %p
+        To attach a Cally calendar to an input in a popover, use the
+        = component_link("Cally Input", "Daisy::DataInput::CallyInputComponent", size: "btn-neutral")
```
Comment (topherfangio) [inline docs/demo/app/views/examples/daisy/data_input/calendars.html.haml:58]: Remove unused size attribute

File: docs/demo/app/views/examples/daisy/data_input/cally_inputs.html.haml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    :markdown
+      The built-in Form Builder extension makes it easy to use Cally Inputs in your
+      Rails forms. The component automatically extracts the name, ID, and value
+      attributes from the form object.
+
+  = form_with(url: "#", method: :post, scope: :event) do |form|
+    .my-2.flex.flex-col.gap-4
+      = form.daisy_cally_input(:start_date, start: "Start Date")
+      = form.daisy_cally_input(:end_date, end: "End Date")
+      = form.daisy_cally_input(:event_date, placeholder: "Select an event date", floating: "Select an event date")
```
Comment (topherfangio) [inline docs/demo/app/views/examples/daisy/data_input/cally_inputs.html.haml:51]: Use floating_placeholder here

File: lib/loco_motion/concerns/labelable_component.rb [hunk 15]
```diff
@@ -58,15 +61,28 @@ module LabelableComponent
       # @option instance_kws [String, nil] :floating Text to display in the
       #   floating label position
       #
+      # @option instance_kws [String, nil] :placeholder The input's placeholder text.
+      #   If not provided and `floating_placeholder` is set, it will use that value.
+      #
+      # @option instance_kws [String, nil] :floating_placeholder Text to use for both
+      #   the floating label and the input placeholder. This is a convenience option
+      #   that sets both the `floating` and `placeholder` options to the same value.
+      #   If both `floating_placeholder` and `floating` are provided, `floating`
```
Comment (topherfangio) [inline lib/loco_motion/concerns/labelable_component.rb:70]: Not sure this is true, remove?

**Review by topherfangio — COMMENTED** [review]
Still need to check specs

File: spec/components/daisy/data_input/cally_component_spec.rb [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+      # Hero icon is likely be stubbed in test mode, so we'll check for the slot attribute
+      # rather than the specific SVG content
+    end
+
+    it "renders custom next icon" do
+      expect(page).to have_css("[slot='next']")
+      # Hero icon is likely be stubbed in test mode, so we'll check for the slot attribute
+      # rather than the specific SVG content
+    end
+  end
```
Comment (topherfangio) [inline spec/components/daisy/data_input/cally_component_spec.rb:125]: These look like bad tests; ideally would check for SVG content.

File: spec/components/daisy/data_input/cally_component_spec.rb [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    it "sets slot attribute for PreviousIcon" do
+      render_inline(previous_icon)
+      expect(page).to have_css("[slot='previous']")
+    end
+
+    it "sets slot attribute for NextIcon" do
+      render_inline(next_icon)
+      expect(page).to have_css("[slot='next']")
+    end
+  end
```
Comment (topherfangio) [inline spec/components/daisy/data_input/cally_component_spec.rb:186]: Again, this should check the right icons...
